articles/150_ros_command_line.md

```diff
@@ -0,0 +1,137 @@
+---
+layout: default
+title: ROS Command Line
+permalink: articles/ros_command_line.html
+abstract:
+  This article describes ROS 2 nodes command line capabilities and its syntax.
+author: '[Michel Hidalgo](https://github.com/hidmic)'
+published: true
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Overview
+
+As it was the case in ROS 1, ROS 2 nodes allow configuration via command line arguments to a certain degree.
+In ROS 2, this interface had to become more complex to cope with a larger set of configuration options, an ambiguity in remapping rules and parameter assignment syntax (as a result of the leading underscore name convention for hidden resources), a one-to-many relationship between executables and nodes, to name a few.
+
+Because of this, increasingly precise addressing mechanisms as well as leading double underscores (`__`) in some positional arguments, both natural extensions of existing ROS 1 command line features, are combined with ROS 2 specific command line flags.
+Flags, in contrast with other custom syntax alternatives, are:
+
+- Widely known and used.
+- Explicit and less error prone.
+- Suitable for tab completion.
+
+Unfortunately, since these flags coexist with user-defined ones, additional guarding and extraction devices must be put in place -- one of the reasons why these were avoided entirely in ROS 1 command lines.
+
+## Features
+
+### Namespacing
+
+To prevent ROS specific command line flags from colliding with user-defined ones, the former are scoped using the `--ros-args` flag and a trailing double dash token (`--`):
+
+```sh
+ros2 run some_package some_node [<user-defined-arg0>...] \
+  --ros-args <ros-specific-arg-0> [<ros-specific-arg-1>...] -- <user-defined-arg-1> [<user-defined-arg-2>...]
```

Comment:
Does `--ros-args` need to be `nargs='+'`? Can't it be defined as `nargs='*'` allowing no ROS specific arguments to follow?

Reply:
Hmm, what use case do you have in mind?

Reply:
Just common logic to build these arguments. If at least one is required the logic needs to double check / special case that when generating the command line.

It is the same for any other argument taking multiple parameters in command line tools, e.g. `colcol build --cmake-args`, `ros2 security generate_artifacts --node-names`.

Reply:
I don't understand what you mean.

Reply:
`ros2 run some_package some_node --ros-args` should be a valid invocation. The current design requires at least one argument (`<ros-specific-arg-0>`) to be present.

Reply:
I understand now, yeah, I think it should be fine to not actually have any ros arguments follow `--ros-args`, so `--ros-args --` would also be ok.